src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1924,24 +1925,34 @@ private void PerformSchedulingActions(IEnumerable<ScheduleResponse> responses)
 
                     case ScheduleActionType.CreateNode:
                         var newNodes = new List<NodeInfo>();
-
-                        for (int i = 0; i < response.NumberOfNodesToCreate; i++)
+                        GetNodeConfiguration();
+                        try
                         {
-                            NodeInfo createdNode = _nodeManager.CreateNode(GetNodeConfiguration(), response.RequiredNodeType);
-
-                            if (null != createdNode)
+                            Parallel.For(0, response.NumberOfNodesToCreate, i =>
                             {
-                                _noNodesActiveEvent.Reset();
-                                _activeNodes.Add(createdNode.NodeId);
-                                newNodes.Add(createdNode);
-                                ErrorUtilities.VerifyThrow(_activeNodes.Count != 0, "Still 0 nodes after asking for a new node.  Build cannot proceed.");
-                            }
-                            else
-                            {
-                                BuildEventContext buildEventContext = new BuildEventContext(0, Scheduler.VirtualNode, BuildEventContext.InvalidProjectInstanceId, BuildEventContext.InvalidProjectContextId, BuildEventContext.InvalidTargetId, BuildEventContext.InvalidTaskId);
-                                ((IBuildComponentHost)this).LoggingService.LogError(buildEventContext, new BuildEventFileInfo(String.Empty), "UnableToCreateNode", response.RequiredNodeType.ToString("G"));
+                                NodeInfo createdNode = _nodeManager.CreateNode(GetNodeConfiguration(), response.RequiredNodeType);
```

Comment:
I don't think `_nodeManager` or the `NodeProviderOutOfProc` it calls into is thread safe, but they're being called in parallel here.

Reply:
Yup, would need to make the whole object tree under `NodeManager` thread safe.